akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -607,19 +607,19 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    * Caution: for n==1 andThen is called after resetting the handler, for
    * other values it is called without resetting the handler.
    */
-  private class Reading[T](in: Inlet[T], private var n: Int, val previous: InHandler)(andThen: T ⇒ Unit, onClose: () ⇒ Unit) extends InHandler {
-    override def onPush(): Unit = {
+  private class Reading[T](in: Inlet[T], private var n: Int, val previous: InHandler)(andThen: T ⇒ Unit, onComplete: () ⇒ Unit) extends InHandler {
+    override def onPush(): Unit = if (n > 0) {
       val elem = grab(in)
-      if (n == 1) setHandler(in, previous)
-      else {
-        n -= 1
-        pull(in)
-      }
+      n -= 1
+
+      if (n > 0) pull(in)
+      else setHandler(in, previous)
+
       andThen(elem)
     }
```

Comment:
shouldn't it set the original handler here back as well? (case when `n==0` in the beginning) Or if that is an invalid parameter to `Reading` then add a `require`.